ApollosProject/apollos-shovel
# Crossroads Kids' Club Backfill (with Contentful)

### About

This PR adds shovel capability for crossroads kids club data coming from Contentful. This PR _only_ includes the backfill.

The steps are:
<img width="664" alt="Screen Shot 2022-01-11 at 12 13 54 PM" src="https://user-images.githubusercontent.com/34075293/148989692-06b3263e-b188-4c54-9a13-f61c1a0de4e4.png">


#### `create_categories`
This is all placeholder code. Categories do not yet exist in Contentful, so we're hard-coding them 😅 I hope this step will be easy to update once kids club categories _are_ available in Contentful.

#### `fetch_and_save_assets`
Contentful calls images `Assets`. In this step, we are fetching all of the assets from Contentful. Yes, even assets not associated with a kids' club Entry are synced. Later, unused assets will be deleted.

#### `fetch_and_save_content_items`
This is the core of the data syncing, and a lot of things are happening:
- we are fetching all `Entries` from contentful
- we find what entries are kids club content, based on their category (and match the required distribution channel) and save those to postgres
- we link the item to its asset
- we save the video URL as a row in the `media` table in postgres
- we link the item to its correct category (older kids, or younger kids, based on the title because these aren't setup in Contentful)
- finally, for a backfill, we delete any items that weren't newly synced

#### `remove_unused_assets`
Here we find all assets that aren't linked to a `content_item` and delete them from the database.

#### So why are you syncing assets first?
For the history of this decision, see the discussion on [this to-do in Basecamp](https://3.basecamp.com/3926363/buckets/25558802/todos/4502039686).

### Testing
Before testing, to airflow, add:
- [ ] A connection to the `crossroads_kids_club_apollos_postgres` (credentials from @vinnyjth)
- [ ] A variable called `crossroads_contentful_api_key` (credentials from @liamillard)

1. Start up Airflow, and filter the dags list by `crossroads`.
2. Unpause the `crossroads_kids_club_backfill_contentful_content_item_dag`. Wait for it to complete, it will take a few minutes.
3. Check out all the data in postgres 😅

### Screenshots
<img width="313" alt="Screen Shot 2022-01-11 at 12 13 47 PM" src="https://user-images.githubusercontent.com/34075293/148989694-c582ef17-6c70-4b20-a1f7-d656989d597e.png">
